cabal-install-solver/src/Distribution/Solver/Types/ConstraintSource.hs

```diff
@@ -1,11 +1,104 @@
 {-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ParallelListComp #-}
 module Distribution.Solver.Types.ConstraintSource
     ( ConstraintSource(..)
+    , RootConfig(..)
+    , Importee(..)
+    , Importer(..)
+    , ImportedConfig(..)
+    , ProjectConfigPath(..)
+    , mkProjectConfigPath
+    , projectConfigPathSource
+    , showProjectConfigPath
     , showConstraintSource
+    , nullProjectConfigPath
     ) where
 
 import Distribution.Solver.Compat.Prelude
 import Prelude ()
+import Data.Coerce (coerce)
+import GHC.Stack (HasCallStack)
+
+-- | Path to the project config file root, typically cabal.project.
+newtype RootConfig = RootConfig FilePath
+    deriving (Eq, Show, Generic)
+
+-- | Path to the project config file with the import.
+newtype Importer = Importer FilePath
+    deriving (Eq, Show, Generic)
+
+-- | Path to the imported file contributing to the project config.
+newtype Importee = Importee FilePath
+    deriving (Eq, Show, Generic)
+
+-- | The imported config along with its import chain.
+data ImportedConfig =
+    ImportedConfig
+        { importers :: [Importer]
+        -- ^ Path to the project config file with the import. Doesn't include the importee.
+        , importee :: Importee
+        -- ^ Path to the imported file contributing to the project config.
+        }
+    deriving (Eq, Show, Generic)
+
+-- | Path to the project config file, either the root or an import.
+data ProjectConfigPath = ProjectRoot RootConfig | ProjectImport ImportedConfig
+    deriving (Eq, Show, Generic)
```

Comment:
I started reviewing the PR, but I had some questions about these types. Why are additional types needed beyond something like this?

```haskell
newtype ProjectConfigPath = ProjectConfigPath [ConfigFilePath]

newtype ConfigFilePath = ConfigFilePath FilePath
```

Why are there three different types of config file paths? Why is a `RootConfig` handled differently than an `Importee`?

Reply:
Thanks @grayjay, I tried using `NonEmpty` and like how it turned out. I'll likely push these changes tomorrow.

Reply:
I'm reviewing the existing cyclical import checks with #9665. Got some debugging to do there. Might be a while ...

Reply:
@grayjay I got the test changes merged ahead of this with #9665 and #9680. I went with `NonEmpty`. Thanks for the suggestion.